Bottom: eafa85605960c59140ba08d65e7da8a08fc638fb
Top:    329759a9a760c9ce249378a64b0c6dbf999468f1
Author: Jason Wang <jasowang@redhat.com>
Date:   2017-08-10 14:29:38 +0800

Refresh of tun-xdp-support

---

diff --git a/drivers/net/tun.c b/drivers/net/tun.c
index 00a3ec7..5892284 100644
--- a/drivers/net/tun.c
+++ b/drivers/net/tun.c
@@ -1281,6 +1281,7 @@ static struct sk_buff *tun_build_skb(struct tun_struct *tun,
 	char *buf;
 	size_t copied;
 	bool xdp_xmit = false;
+	int err;
 
 	if (unlikely(!skb_page_frag_refill(buflen, alloc_frag, GFP_KERNEL)))
 		return ERR_PTR(-ENOMEM);
@@ -1311,6 +1312,13 @@ static struct sk_buff *tun_build_skb(struct tun_struct *tun,
 		act = bpf_prog_run_xdp(xdp_prog, &xdp);
 
 		switch (act) {
+		case XDP_REDIRECT:
+			get_page(alloc_frag->page);
+			alloc_frag->offset += buflen;
+			err = xdp_do_redirect(tun->dev, &xdp, xdp_prog);
+			if (err)
+				goto err_redirect;
+			return NULL;
 		case XDP_TX:
 			xdp_xmit = true;
 			/* fall through */
@@ -1350,6 +1358,8 @@ static struct sk_buff *tun_build_skb(struct tun_struct *tun,
 
 	return skb;
 
+err_redirect:
+	put_page(alloc_frag->page);
 err_xdp:
 	rcu_read_unlock();
 	this_cpu_inc(tun->pcpu_stats->rx_dropped);
